### ISSUE TITLE
Task should be an interface, not an anonymous function

### ISSUE THREAD

**Issue by nathanjordan** [shicky/gocelery#12]
```
type Task interface {
    Execute() (result interface{}, err error)
}
```

This way you can add dependencies without accessing global state.

For example

```
type AddTask struct {
    A `json:"a"`
    B `json:"b"`
    Cache CachingService
}
```

**Comment by shicky** [comment]
Thank you for the issue.
I think its great idea but it might also add burden of extra code to users.
You can achieve same effect by registering member functions instead.
Take a look at below example.

```go
type MyObject struct {
	MyInt int
}

func (mo *MyObject) add(a int, b int) int {
	return a + b + mo.MyInt
}

func main() {
	mo := &MyObject{999}
	celeryBroker := gocelery.NewRedisCeleryBroker("localhost:6379", "")
	celeryBackend := gocelery.NewRedisCeleryBackend("localhost:6379", "")
	celeryClient, _ := gocelery.NewCeleryClient(celeryBroker, celeryBackend, 2)
	celeryClient.Register("worker.add", mo.add)
	go celeryClient.StartWorker()
	time.Sleep(30 * time.Second)
	celeryClient.StopWorker()
}

```

**Comment by nathanjordan** [comment]
Even if you _CAN_ pass a bound method to `Register` (does it need to have a pointer receiver to do this?), in idiomatic Go this kind of thing is usually accomplished using an interface. Most use cases for Celery I've experienced in production environments involve a lot of moving pieces and are rarely very simple functions like this. 

If you don't want to change the API and still have this as a feature, a runtime type assertion could determine if the second parameter of `Register` implements the `Task` interface and can handle it differently.

**Comment by shicky** [comment]
Above code works and [pointer to interface is not correct type to hold pointer to struct.](http://stackoverflow.com/a/13511853)

I assume that you want to model after [Celery task class](http://docs.celeryproject.org/en/latest/userguide/tasks.html#instantiation) in Python implementation.

Certainly your approach would mirror that.
You are always welcome to create pull request =)

**Comment by nathanjordan** [comment]
I think you're misunderstanding what I'm trying to say. The article you linked is about pointers to interfaces. I'm referring to the method you're adding to register (`add`) having a [pointer receiver](https://golang.org/doc/faq#methods_on_values_or_pointers).

Regardless, I think a PR will clear up confusion. The "Celery task class" is the correct approach to take in Go.

**Comment by shicky** [comment]
Oh I see what you mean.
Yes I guess there are cases where that is needed.

**Comment by shicky** [comment]
Using interface would provide performance gain over using reflection.
Only arguments passed through `kwargs` dictionary is supported using `CeleryTask` interface.

merged at a14372d12fc85b6a39ed3ce79f142435a38660e0
